File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.cpp
class Solution {
public:
    
    void inorder(TreeNode* root,vector<int>& answer){
        if(!root) return;
        inorder(root->left, answer);
        answer.push_back(root->val);
        inorder(root->right, answer);
    }
    
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int>answer;
        inorder(root,answer);
        return answer;
    }
};

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.java
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> answer = new ArrayList<>();
        while(root != null){
            if(root.left == null){
                answer.add(root.val);
                root = root.right;
            }
            else{
                TreeNode leftNode = root.left;
                while(leftNode != null && leftNode.right != null && leftNode.right != root){
                    leftNode = leftNode.right;
                }
                if(leftNode.right == null){
                    leftNode.right = root;
                    root = root.left;
                }
                else{
                    answer.add(root.val);
                    leftNode.right = null;
                    root = root.right;
                }
            }
        }
        return answer;
    }
}